api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,209 @@
+package org.openmrs.module.pihcore.event;
+
+import org.apache.commons.dbutils.handlers.ScalarHandler;
+import org.apache.commons.lang.BooleanUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.openmrs.module.dbevent.Database;
+import org.openmrs.module.dbevent.DatabaseTable;
+import org.openmrs.module.dbevent.DbEvent;
+import org.openmrs.module.dbevent.DbEventSourceConfig;
+import org.openmrs.module.dbevent.EventConsumer;
+import org.openmrs.module.dbevent.Operation;
+import org.openmrs.module.dbevent.Rocks;
+import org.openmrs.module.dbevent.SqlBuilder;
+
+import java.io.File;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Consumes patient-related events
+ * For performance reasons on larger databases, this does not use a Debezium-generated snapshot, but rather
+ * computes a custom snapshot that is then updated as needed from change events.
+ * When used, this should be configured with "snapshot.mode = schema_only"
+ */
+public class PatientUpdateEventConsumer implements EventConsumer {
+
+    private static final Logger log = LogManager.getLogger(PatientUpdateEventConsumer.class);
+
+    private final DbEventSourceConfig config;
+    private final Database database;
+    private final Map<String, String> patientKeys = new LinkedHashMap<>();
+    private Rocks statusDb;
+    private boolean snapshotInitialized = false;
+
+    public PatientUpdateEventConsumer(DbEventSourceConfig config) {
+        this.config = config;
+        this.database = config.getContext().getDatabase();
+        patientKeys.put("patient_id", "patient");
+        patientKeys.put("person_id", "person");
+        patientKeys.put("person_a", "person");
+        patientKeys.put("person_b", "person");
+        patientKeys.put("order_id", "orders");
+        patientKeys.put("patient_program_id", "patient_program");
+        patientKeys.put("encounter_id", "encounter");
+        patientKeys.put("visit_id", "visit");
+        patientKeys.put("allergy_id", "allergy");
+        patientKeys.put("appointment_id", "appointmentscheduling_appointment");
+    }
+
+    @Override
+    public void startup() {
+        try {
+            statusDb = new Rocks(new File(config.getContext().getModuleDataDir(), "status.db"));
+            snapshotInitialized = BooleanUtils.isTrue(statusDb.get("snapshotInitialized"));
+        }
+        catch (Exception e) {
+            throw new RuntimeException("An error occurred starting up", e);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        statusDb.close();
+    }
+
+    @Override
+    public void accept(DbEvent event) {
+        if (!snapshotInitialized) {
+            performInitialSnapshot();
```

Comment:
I recently moved the initialization code in here, as this ensures that it doesn't run until after the Debezium engine has started up and such.  I previously had it in the startup method.  We may want to put it back there (it fits there more naturally), I only moved it here as I was debugging some issues and thought this might be the cause.  The problem with putting it here is that it won't execute until the first event comes in.  So if the server is fired up in the middle of the night, this will be idle until someone (or a scheduled task) comes along and saves some sort of patient data, at which point this method will get triggered, it'll see that it needs to be initialized, and it'll do so, which will take some minutes to perform on a large DB given it's querying every patient table (including every Obs).